### ISSUE TITLE
Block Theme Previews does not work if we call the `get_template` or `get_stylesheet` functions early

### ISSUE THREAD

**Issue by okmttdhr** [WordPress/gutenberg#53284]
### Description

Block Theme Previews does not work if we call the `get_template` or `get_stylesheet` functions before the pluggable functions are not ready. So, if a 3rd party plugin (e.g. CoBlocks, Elementor, Classic Editor, and WooCommerce) needs to access the template or stylesheet before the pluggable functions are ready, then the site with such a plugin will break (see the reproduction instruction below).

The reason this error happens is;

- Block Theme Previews adds hooks to `stylesheet` and `template`. https://github.com/WordPress/wordpress-develop/blob/4417367/src/wp-includes/theme-previews.php#L80-L81
- It checks `current_user_can( 'switch_themes' )`, which uses `wp_get_current_user` inside of it. https://github.com/WordPress/wordpress-develop/blob/4417367/src/wp-includes/theme-previews.php#L17-L19
- `wp_get_current_user` is initialized after `pluggable.php` is called. https://github.com/WordPress/wordpress-develop/blob/1d7554d/src/wp-settings.php#L482-L484

A possible solution would be one of the following;

- Load the [pluggable functions](https://github.com/WordPress/wordpress-develop/blob/1d7554d/src/wp-settings.php#L482-L484) early
- Load [`theme-previews.php`](https://github.com/WordPress/wordpress-develop/blob/1d7554d/src/wp-settings.php#L184) after the pluggable functions
- Load the pluggable functions on demand in `theme-previews.php`
- Tell developers for the 3rd party plugins not to access the template or stylesheet directly when loading plugins (a developer guide should be covered, maybe)
- Something else

### Step-by-step reproduction instructions

1. Add a file in `wp-content/plugins/sample-plugin/sample-plugin.php` on your site.

```php
<?php
/*
Plugin Name: Sample plugin
*/

$wp_theme = wp_get_theme();
$wp_theme->get_stylesheet();
```

2. Go to `/wp-admin/themes.php` and click `Live Preview` button on your installed **Block theme**. 
3. See the error: `Uncaught Error: Call to undefined function wp_get_current_user()`

### Screenshots, screen recording, code snippet

<img width="804" alt="Screen Shot 2023-08-03 at 16 25 19" src="https://github.com/WordPress/gutenberg/assets/5287479/feb12b5f-60fe-4db9-a9d1-43bb5c47991f">

### Environment info

- WordPress: 6.3-RC3-56344
- Gutenberg: 16.3.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by scruffian** [comment]
I had an idea for how to solve this here: https://github.com/WordPress/wordpress-develop/pull/4959

**Comment by okmttdhr** [comment]
> So, if a 3rd party plugin needs to access the template or stylesheet before the pluggable functions are ready, then the site with such a plugin will break

For example, it breaks if a user has the [CoBlocks](https://wordpress.org/plugins/coblocks/) plugin activated.
https://github.com/godaddy-wordpress/coblocks/blob/1bf29ec/class-coblocks.php#L225

**Update**: This issue happens on the site with a plugin like Elementor, and WooCommerce.

**Comment by okmttdhr** [comment]
> I had an idea for how to solve this here: [WordPress/wordpress-develop#4959](https://github.com/WordPress/wordpress-develop/pull/4959)

Thanks for trying this out, @scruffian!
However, I've noticed some comments on https://core.trac.wordpress.org/ticket/59000 expressing concerns about its reliability.

>Load [theme-previews.php](https://github.com/WordPress/wordpress-develop/blob/1d7554d/src/wp-settings.php?rgh-link-date=2023-08-03T07%3A31%3A55Z#L184) after the pluggable functions

This alternative seems more fitting to me, as I feel that performing capability checks before defining pluggable functions itself might not be the optimal approach. What are your thoughts?

**Comment by fushar** [comment]
@okmttdhr @scruffian, this has been shipped and backported to WP 6.3.2, should we close this issue?

**Comment by okmttdhr** [comment]
Closing, as this issue has been solved at https://core.trac.wordpress.org/ticket/59000!
